File: src/main/java/kookmin/software/capstone2023/timebank/application/service/auth/DefaultUserAuthenticator.java
package kookmin.software.capstone2023.timebank.application.service.auth;

import kookmin.software.capstone2023.timebank.application.exception.UnauthorizedException;
import kookmin.software.capstone2023.timebank.application.service.auth.token.AccessTokenService;
import kookmin.software.capstone2023.timebank.domain.repository.AccountJpaRepository;
import kookmin.software.capstone2023.timebank.domain.repository.UserJpaRepository;

public class DefaultUserAuthenticator implements UserAuthenticator {
    private final AccessTokenService accessTokenService;
    private final UserJpaRepository userJpaRepository;
    private final AccountJpaRepository accountJpaRepository;

    public DefaultUserAuthenticator(
            AccessTokenService accessTokenService,
            UserJpaRepository userJpaRepository,
            AccountJpaRepository accountJpaRepository
    ) {
        this.accessTokenService = accessTokenService;
        this.userJpaRepository = userJpaRepository;
        this.accountJpaRepository = accountJpaRepository;
    }

    @Override
    public AuthenticationData authenticate(String accessToken) {
        AccessTokenService.AccessTokenClaims claims = accessTokenService.verify(accessToken);

        User user = userJpaRepository.findById(claims.getUserId()).value;
        if (user == null) {
            throw new UnauthorizedException(null);
        }

        Account account = accountJpaRepository.findById(claims.getAccountId()).value;
        if (account == null) {
            throw new UnauthorizedException(null);
        }

        if (user.getAccount().getId() != account.getId()) {
            throw new UnauthorizedException(null);
        }

        return new AuthenticationData(user.getId(), account.getId(), account.getType());
    }
}

File: src/main/java/kookmin/software/capstone2023/timebank/application/service/auth/DefaultUserAuthenticator.kt
package kookmin.software.capstone2023.timebank.application.service.auth

import kookmin.software.capstone2023.timebank.application.exception.UnauthorizedException
import kookmin.software.capstone2023.timebank.application.service.auth.token.AccessTokenService
import kookmin.software.capstone2023.timebank.domain.repository.AccountJpaRepository
import kookmin.software.capstone2023.timebank.domain.repository.UserJpaRepository
import org.springframework.data.repository.findByIdOrNull

class DefaultUserAuthenticator(
    private val accessTokenService: AccessTokenService,
    private val userJpaRepository: UserJpaRepository,
    private val accountJpaRepository: AccountJpaRepository,
) : UserAuthenticator {
    override fun authenticate(accessToken: String): UserAuthenticator.AuthenticationData {
        val claims = accessTokenService.verify(accessToken)

        val user = userJpaRepository.findByIdOrNull(claims.userId)
            ?: throw UnauthorizedException(message = null)

        val account = accountJpaRepository.findByIdOrNull(claims.accountId)
            ?: throw UnauthorizedException(message = null)

        if (user.account.id != account.id) {
            throw UnauthorizedException(message = null)
        }

        return UserAuthenticator.AuthenticationData(
            userId = user.id,
            accountId = account.id,
            accountType = account.type,
        )
    }
}
